keps/sig-instrumentation/2831-kubelet-tracing/README.md

```diff
@@ -0,0 +1,366 @@
+* Replace existing logging, metrics
+* Change metrics or logging (e.g. to support trace-metric correlation)
+* Access control to tracing backends
+* Add tracing to components outside kubernetes (e.g. etcd client library).
+
+## Proposal
+
+### User Stories
+
+Since this feature is for diagnosing problems with the kubelet, it is targeted at cluster operators, administrators, and cloud vendors that 
+manage kubernetes nodes and core services.
+
+For the following use-cases, I can deploy an OpenTelemetry collector agent as a DaemonSet to collect kubelet trace data from each node's host network.
+From there, OpenTelemetry trace data can be exported to a tracing backend of choice.  I can configure the kubelet's optional `TracingConfiguration`
+specified within `kubernetes/component-base` with the default URL to make the kubelet send its spans to the collector.
+Alternatively, I can point the kubelet at an OpenTelemetry collector listening on a different port or URL if I need to.
+
+#### Continuous Trace Collection
+
+As a cluster administrator or cloud provider, I would like to collect gRPC and HTTP trace data from the transactions between the API server and the 
+kubelet and interactions with a node's container runtime (Container Runtime Interface) to debug cluster problems.  I can set the `SamplingRatePerMillion`
+in the configuration file to a non-zero number to have spans collected for a small fraction of requests. Depending on the symptoms I need to
+debug, I can search span metadata or specific nodes to find a trace which displays the symptoms I am looking to debug.
+The sampling rate for trace exports can be configured based on my needs. I can collect each node's kubelet trace data as distinct tracing services
+to diagnose node issues.
+
+##### Example Scenarios
+
+* Latency or timeout experienced when:
+    * Attach or exec to running containers
+        * APIServer to Kubelet, then Kubelet to CRI can determine which component/operation is the culprit
+    * Kubelet's port-forwarding functionality
+        * APIServer, Kubelet, CRI trace data can determine which component/operation is experiencing issues
+    * Container Start/Stop/Create/Remove
+        * Kubelet, CRI trace data determines which component/operation is experiencing issues with insight into which is experiencing issues.
+* Kubelet spans can detect node-level latency - each node produces trace data filterable with node hostname
+    * Are there issues with particular regions? Cloud providers? Machine-types?
+
+### Tracing Requests and Exporting Spans
+
+The gRPC client in the kubelet's Container Runtime Interface (CRI) [Remote Runtime Service](https://github.com/kubernetes/kubernetes/blob/release-1.21/pkg/kubelet/cri/remote/remote_runtime.go) and [the CRI streaming package](https://github.com/kubernetes/kubernetes/tree/release-1.21/pkg/kubelet/cri/streaming) will be instrumented to export trace data and propagate trace context.
```

Comment:
Will it be possible to enumerate an initial set of OpenTelemetry Attributes for the traces that will be exported by the Kubelet (either here or perhaps in the Design Details section below)?

Reply:
these details can be added as follow-ups, i think. once this is approved we'll get more specific and can add a follow-up to update the proposal, esp if you have a good idea for what you'd want in the attributes : )